### PR TITLE
Light-weight benchmarking script

### PR DESCRIPTION
**Pull request by NusretOzates**
Implemented sentiment analysis benchmark for Classifiers using IMDB review dataset for #634

### CODE REVIEW

**Comment by chenmoneygithub** [comment]
@NusretOzates Thank you for the PR! Will take a closer look tomorrow.

File: keras_nlp/benchmarks/README.md [hunk 1]
```diff
@@ -25,3 +25,28 @@ the following results were obtained:
 To change the configuration, say, for example, number of layers in the transformer
 model used for inference, the user can modify the config dictionaries given at
 the top of the script.
+
```
Comment (mattdangerw) [inline keras_nlp/benchmarks/README.md:28]: only one blank line before and after headings

File: keras_nlp/benchmarks/sentiment_analysis.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        .prefetch(tf.data.AUTOTUNE)
+    )
+
+    return train_dataset, val_dataset, test_dataset
+
+
+def create_model():
+    for name, symbol in keras_nlp.models.__dict__.items():
+        if inspect.isclass(symbol) and issubclass(symbol, keras.Model):
+            if FLAGS.model and name != f"{FLAGS.model.capitalize()}Classifier":
```
Comment (mattdangerw) [inline keras_nlp/benchmarks/sentiment_analysis.py:78]: Rather than all this, I would just take in the symbol name directly e.g. `--model=BertClassifier`. This will be a little more obvious in usage.

File: keras_nlp/benchmarks/sentiment_analysis.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    dataset = tfds.load("imdb_reviews", as_supervised=True)
+    train_dataset, test_dataset = dataset["train"], dataset["test"]
+
+    train_dataset = (
+        train_dataset.shuffle(BUFFER_SIZE)
+        .batch(FLAGS.batch_size)
+        .prefetch(tf.data.AUTOTUNE)
+    )
+    val_dataset = (
+        test_dataset.take(10000)
```
Comment (mattdangerw) [inline keras_nlp/benchmarks/sentiment_analysis.py:62]: Maybe rather than a hardcoded split like this, you could run [tfds.load](https://www.tensorflow.org/datasets/api_docs/python/tfds/load) `with_info`. Use that to get the size of the test set, and use a fractional split here. E.g. `int(test_dataset_cardinality / 2)`

File: keras_nlp/benchmarks/sentiment_analysis.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    check_flags()
+    train_dataset, validation_dataset, test_dataset = create_imdb_dataset()
+    model = create_model()
+    model = train_model(model, train_dataset, validation_dataset)
+    evaluate_model(model, test_dataset)
+
+    # End time
+    end_time = time.time()
+    print(f"Total time: {end_time - start_time}")
```
Comment (mattdangerw) [inline keras_nlp/benchmarks/sentiment_analysis.py:132]: Maybe "Wall time", so it's clear we are just measuring elapsed time?

File: keras_nlp/benchmarks/README.md [hunk 5]
```diff
@@ -25,3 +25,28 @@ the following results were obtained:
...
+```
+
+flag `--model` specifies the model name, and `--preset` specifies the preset under testing. `--preset` could be None, 
+while `--model` is required. Other flags are common training flags.
+
+This script outputs:
+
+- validation accuracy for each epoch.
+- testing accuracy after training is done.
+- total elapsed time (in seconds).
```
Comment (mattdangerw) [inline keras_nlp/benchmarks/README.md:52]: Will this display in some way the average train step time during fit? I think that is an important metric for us.

**Review by mattdangerw — COMMENTED** [review]
LGTM! Just a few comments.

File: keras_nlp/benchmarks/sentiment_analysis.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
+# Copyright 2022 The KerasNLP Authors
```
Comment (chenmoneygithub) [inline keras_nlp/benchmarks/sentiment_analysis.py:1]: We can use 2023 now, time flashes!

File: keras_nlp/benchmarks/sentiment_analysis.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+from tensorflow import keras
+
+import keras_nlp
+
+# Use mixed precision for optimal performance
+keras.mixed_precision.set_global_policy("mixed_float16")
+
+FLAGS = flags.FLAGS
+flags.DEFINE_string(
+    "model", None, "The name of the classifier such as BertClassifier."
```
Comment (chenmoneygithub) [inline keras_nlp/benchmarks/sentiment_analysis.py:30]: Add a training comma so that it's formatted to multilines.

File: keras_nlp/benchmarks/sentiment_analysis.py (hunk 3 above)
Comment (chenmoneygithub) [reply keras_nlp/benchmarks/sentiment_analysis.py:62]: I think you can do:

```
test_ds_size = test_dataset.cardinality()
val_dataset = test_dataset.take(test_ds_size // 2)
...
```

File: keras_nlp/benchmarks/sentiment_analysis.py (hunk 2 above)
Comment (chenmoneygithub) [reply keras_nlp/benchmarks/sentiment_analysis.py:78]: Yea, this is my bad honestly, I put "bert" in the problem description.

**Review by chenmoneygithub — CHANGES_REQUESTED** [review]
Looks great overall! A few comments.

**Comment by NusretOzates** [comment]
Will do all of that tomorrow! Thanks for the review @chenmoneygithub @mattdangerw

File: keras_nlp/benchmarks/README.md (hunk 5 above)
Comment (NusretOzates) [reply keras_nlp/benchmarks/README.md:52]: I think I can do that by creating a Callback subclass. Do you have a more straightforward suggestion?

Comment (mattdangerw) [reply keras_nlp/benchmarks/README.md:52]: I'm not sure actually! Let me get back to you.

I know we would like to have something indicative of the time elapsed for only the forward and backwards pass. The total elapsed time will also include compilation (which is good too, but not always the metrics we will want).

Comment (NusretOzates) [reply keras_nlp/benchmarks/README.md:52]: Actually, how about this? Btw After the "Compiler cluster using XLA" line, I had to wait more than 10 mins. I don't know the reason maybe it is because of WSL2

image

Comment (mattdangerw) [reply keras_nlp/benchmarks/README.md:52]: Yeah, I think the train step time printed by model fit is totally fine at least for now.

Maybe as a follow up we should add a flag to enable tensorboard and the tensorboard profiler. https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras

**Comment by NusretOzates** [comment]
@chenmoneygithub @mattdangerw, I made the necessary changes. I would like to add a TensorFlow profiler too but tensorboard and tensorboard_profiler_plugin are not in the requirements and I didn't want to touch that part 😄

File: keras_nlp/benchmarks/sentiment_analysis.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+flags.DEFINE_integer("num_epochs", 1, "The number of epochs.")
+flags.DEFINE_integer("batch_size", 16, "The batch size.")
+
+tfds.disable_progress_bar()
+
+BUFFER_SIZE = 10000
+
+
+def check_flags():
+    if not FLAGS.model:
```
Comment (mattdangerw) [inline keras_nlp/benchmarks/sentiment_analysis.py:50]: I just remembered there is actually a way to do this with absl directly. `flags.mark_flag_as_required("flag")`.

https://github.com/keras-team/keras-nlp/blob/master/examples/bert_pretraining/bert_pretrain.py#L454

File: keras_nlp/benchmarks/sentiment_analysis.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from absl import app
+from absl import flags
+from tensorflow import keras
+
+import keras_nlp
+
+# Use mixed precision for optimal performance
+keras.mixed_precision.set_global_policy("mixed_float16")
```
Comment (mattdangerw) [inline keras_nlp/benchmarks/sentiment_analysis.py:26]: Actually given that this is a benchmarking script, better to leave this as a flag probably. Totally valid to benchmark a model under full precision or mixed.

Can we just make this a string flag?

```
flags.DEFINE_string(
    "mixed_precision_policy",
    "mixed_float16",
    "The global mixed precision policy to use. E.g. 'mixed_float16' or 'float32'.",
)
```

File: keras_nlp/benchmarks/sentiment_analysis.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    )
+    train_dataset, test_dataset = dataset["train"], dataset["test"]
+
+    train_dataset = (
+        train_dataset.shuffle(BUFFER_SIZE)
+        .batch(FLAGS.batch_size)
+        .prefetch(tf.data.AUTOTUNE)
+    )
+
+    test_dataset_size = info.splits['test'].num_examples // 2
```
Comment (mattdangerw) [inline keras_nlp/benchmarks/sentiment_analysis.py:66]: Maybe drop a comment here. 

`# We split the test data evenly into validation and test sets.`

**Review by mattdangerw — APPROVED** [review]
LGTM! Just a few last comments, but after those are in I think this is all set to merge.

Agreed the tensorboard profiler bit can be a follow up.

Thanks so much for this! This is high quality code.

**Comment by NusretOzates** [comment]
@mattdangerw All done👌 I'm happy to contribute and I actually would like/plan to do more!

**Comment by mattdangerw** [comment]
@NusretOzates thanks! I see one small issue, we now need to make sure to only set the mixed precision policy after flags are parsed, or the script wont run. I will push a small fix.

**Comment by NusretOzates** [comment]
That makes sense, thanks for the fix 😄

**Review by chenmoneygithub — APPROVED** [review]
The code looks great! We want to thank you for the high-quality PR again! XD

**Comment by NusretOzates** [comment]
You are welcome! 😄
